pkg/vsphere/vsan/synced_cache.go

```diff
@@ -0,0 +1,439 @@
+			ret = append(ret, requests[i])
+		}
+	}
+	return ret
+}
+
+// doRefresh should be called by run only to make sure concurrency is handled properly
+func (c *syncedDomCache) doRefresh(ctx context.Context, requests []*dsWaitGroup) error {
+	defer trace.End(trace.Begin(fmt.Sprintf("%#v", requests)))
+	dsws := c.getNonRefreshingDS(requests)
+	if len(dsws) == 0 {
+		log.Debugf("Requested datastores %#v is refreshing", requests)
+		return nil
+	}
+
+	c.m.Lock()
+	defer c.m.Unlock()
+	var errs []string
+	for i, dsw := range dsws {
+		dsc, ok := c.dsMap[dsw.dsRef]
+		if !ok {
+			err := errors.Errorf("requested datastore %s does not exist", c.dsMap[dsw.dsRef])
+			log.Error(err)
+			errs = append(errs, err.Error())
+			continue
+		}
+		if dsws[i].wg != nil {
+			c.refreshings[dsw.dsRef] = append(c.refreshings[dsw.dsRef], dsws[i].wg)
+			log.Debugf("waiting groups for datastore %s: %#v", dsw.dsRef, c.refreshings[dsw.dsRef])
+		} else {
+			log.Debugf("wait group for datastore %s is empty", dsws[i].dsRef)
+			if _, ok := c.refreshings[dsw.dsRef]; !ok {
+				// no refreshing item yet
+				c.refreshings[dsw.dsRef] = nil
+			}
+		}
+		go func(dsRef string) {
+			if err := dsc.Refresh(ctx); err != nil {
+				log.Error(err)
+			}
+			c.doneOnce <- dsRef
+		}(dsw.dsRef)
+	}
+	if len(errs) > 0 {
+		return errors.Errorf(strings.Join(errs, "\n"))
+	}
+	return nil
+}
+
+// Start cache refresh after first refresh request coming. After that, refresh once every 5 minutes.
+// Allow to inject more refresh request between timer timeout. But if the request comes while refresh is running, append the refresh waitgroup only, without new refresh triggered
+func (c *syncedDomCache) run() {
+	var tick <-chan time.Time
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for {
+		select {
+		case r := <-c.refresh:
+			log.Debugf("start refresh datastore %s", r.dsRef)
+			c.doRefresh(ctx, []*dsWaitGroup{r})
+		case <-c.stop:
+			log.Debugf("Stop vsan dom cache refresh")
+			c.closeChannels()
+			return
+		case ds := <-c.doneOnce:
+			log.Debugf("datastore %s is refreshed", ds)
+			func() {
+				c.m.Lock()
+				defer c.m.Unlock()
+
+				if _, ok := c.refreshings[ds]; ok {
+					c.doneWaitGroups(c.refreshings[ds])
+				}
+				delete(c.refreshings, ds)
+				if len(c.refreshings) == 0 {
+					log.Debugf("All refresh finished, restart timer")
+					tick = time.After(refreshInterval)
+				}
+			}()
+		case w := <-c.waitCurrentRefresh:
+			log.Debug("waiting refresh for datastore %s", w.dsRef)
+			func() {
+				c.m.RLock()
+				defer c.m.RUnlock()
+				if _, ok := c.refreshings[w.dsRef]; ok {
+					if w.wg != nil {
+						c.refreshings[w.dsRef] = append(c.refreshings[w.dsRef], w.wg)
+					} else {
+						log.Errorf("waiting request for datastore %s should have non-nil waitgroup.", w.dsRef)
+					}
+				} else {
+					log.Debugf("No refresh running for datastore %s, done waitgroup immediately", w.dsRef)
+					w.wg.Done()
+				}
+			}()
+		case <-tick:
```

Comment:
if timer timeout, it will refresh all datastores. But Refresh() method will trigger refresh for one particular datastore